File: Graphs/lee algorithm/lee.py
from collections import deque

def lee_algorithm(grid, start, end):
    rows, cols = len(grid), len(grid[0])
    directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]  # 4-connected neighborhood

    queue = deque([(start[0], start[1], 0)])  # (x, y, distance)
    visited = set(start)

    while queue:
        x, y, distance = queue.popleft()

        if (x, y) == end:
            return distance

        for dx, dy in directions:
            nx, ny = x + dx, y + dy

            if 0 <= nx < rows and 0 <= ny < cols and grid[nx][ny] == 0 and (nx, ny) not in visited:
                queue.append((nx, ny, distance + 1))
                visited.add((nx, ny))

    return -1  # No path found

# Example usage:
grid = [
    [0, 1, 0, 0, 0],
    [0, 1, 0, 1, 0],
    [0, 0, 0, 0, 0],
    [0, 1, 1, 1, 0],
    [0, 0, 0, 1, 0]
]
start = (0, 0)
end = (4, 4)

shortest_distance = lee_algorithm(grid, start, end)
if shortest_distance != -1:
    print(f"Shortest distance from {start} to {end} is {shortest_distance}")
else:
    print("No path found")

File: Graphs/lee algorithm/lee.js
function leeAlgorithm(grid, start, end) {
  const rows = grid.length;
  const cols = grid[0].length;
  const directions = [
    [1, 0],
    [-1, 0],
    [0, 1],
    [0, -1], // 4-connected neighborhood
  ];

  const queue = [[start[0], start[1], 0]]; // [x, y, distance]
  const visited = new Set([`${start[0]},${start[1]}`]);

  while (queue.length > 0) {
    const [x, y, distance] = queue.shift();

    if (x === end[0] && y === end[1]) {
      return distance;
    }

    for (const [dx, dy] of directions) {
      const nx = x + dx;
      const ny = y + dy;

      if (
        nx >= 0 &&
        nx < rows &&
        ny >= 0 &&
        ny < cols &&
        grid[nx][ny] === 0 &&
        !visited.has(`${nx},${ny}`)
      ) {
        queue.push([nx, ny, distance + 1]);
        visited.add(`${nx},${ny}`);
      }
    }
  }

  return -1; // No path found
}

// Example usage:
const grid = [
  [0, 1, 0, 0, 0],
  [0, 1, 0, 1, 0],
  [0, 0, 0, 0, 0],
  [0, 1, 1, 1, 0],
  [0, 0, 0, 1, 0],
];
const start = [0, 0];
const end = [4, 4];

const shortestDistance = leeAlgorithm(grid, start, end);
if (shortestDistance !== -1) {
  console.log(
    `Shortest distance from ${start} to ${end} is ${shortestDistance}`
  );
} else {
  console.log("No path found");
}
